### PR TITLE
Enumerable methods

### PR DESCRIPTION
**Pull request by collinsugwu**
**### What this PR does**

1. creates an Enumerable module
2. write a custom each method
3. use the custom each method to perform operations
4. use proc and yields

### CODE REVIEW

File: enumerable_methods.rb [hunk 1]
```diff
@@ -0,0 +1,71 @@
+# frozen_string_literal: true
+
+module Enumerable
+  def my_each
+    i = 0
+    while i < length
+      yield self[i]
+      i += 1
+    end
+  end
```
Comment (wathigo) [inline enumerable_methods.rb]: **[REQUIRED]** Returns an Enumerator if no block is given.

File: enumerable_methods.rb [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+  def my_each_with_index
+    i = 0
+    while i < length
+      yield self[i], i
+      i += 1
+    end
+  end
```
Comment (wathigo) [inline enumerable_methods.rb]: **[REQUIRED]** Returns an enumerator if no block is given

File: enumerable_methods.rb [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+  def my_select
+    result = []
+    my_each { |i| result << i if yield(i) }
+    result
+  end
```
Comment (wathigo) [inline enumerable_methods.rb]: **[REQUIRED]** Returns an enumerator if no block is given

File: enumerable_methods.rb [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+  def my_any?
+    my_each { |i| true if yield(i) }
+  end
+
+  def my_all?
+    my_each { |i| true if yield(i) == true }
+  end
```
Comment (wathigo) [inline enumerable_methods.rb]: **[REQUIRED]**
* Returns true if the block never returns false or nil.
* When no block or argument is given, it returns true when none of the collection members are false or nil
* When a class is passed as an argument, it returns true if all of the collection is a member of such class.
* When a Regex is passed as an argument, it returns true if all of the collection matches the Regex.
* When a pattern other than Regex or a Class is given, it returns true if all of the collection matches the pattern.

File: enumerable_methods.rb [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+  def my_select
+    result = []
+    my_each { |i| result << i if yield(i) }
+    result
+  end
+
+  def my_any?
+    my_each { |i| true if yield(i) }
+  end
```
Comment (wathigo) [inline enumerable_methods.rb]: **[REQUIRED]**
* Returns true if the block ever returns a value other than false or nil.
* When no block or argument is given, it returns true if at least one of the collection is not false or nil.
* When a class is passed as an argument, it returns true if at least one of the collection is a member of such class.
* When a Regex is passed as an argument, it returns false if none of the collection matches the Regex
* When a pattern other than Regex or a Class is given, it returns false if none of the collection matches the pattern.

File: enumerable_methods.rb [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+  def my_all?
+    my_each { |i| true if yield(i) == true }
+  end
+
+  def my_none?
+    true unless block_given?
+    result = false
+    my_each { |i| true unless yield(i) }
+    result
+  end
```
Comment (wathigo) [inline enumerable_methods.rb]: **[REQUIRED]**
* Returns true if the block never returns true for all elements.
* When no block or argument is given, it returns true only if none of the collection members is true.
* When a class is passed as an argument, it returns true if none of the collection is a member of such class.
* When a Regex is passed as an argument returns true only if none of the collection matches the Regex.
* When a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern.

File: enumerable_methods.rb [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+  def my_count
+    i = 0
+    my_each { i += 1 }
+    i
+  end
```
Comment (wathigo) [inline enumerable_methods.rb]: **[REQUIRED]**
* Counts the number of items in enum that are equal to item if an argument is given.
* Counts the number of elements yielding a true value if a block is given.

File: enumerable_methods.rb [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+  def my_map(&my_proc)
+    result = []
+    my_each { |element| result << my_proc.call(element) }
+    result
+  end
```
Comment (wathigo) [inline enumerable_methods.rb]: **[REQUIRED]**
* Returns an Enumerator if no block is given.

File: enumerable_methods.rb [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+
+  def my_inject(acc = nil)
+    my_each { |i| acc = yield(acc, i) }
+    acc
+  end
+end
```
Comment (wathigo) [inline enumerable_methods.rb]: **[REQUIRED]**
* Combines all elements of enum by applying a binary operation, specified by a block.
* When a symbol is specified, it combines each element of the collection by applying the symbol as a named method

**Review by wathigo — CHANGES_REQUESTED** [review]
### Hi @collinsugwu  :wave:
Great work on the project!
Let's fix a couple of this before we move on.

See the comments below :point_down:  

**After implementing the changes, submit another request for review.**

_Happy Coding :computer: 
Reviewed by: **Simon Wathigo**_

File: enumerable_methods.rb [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+
+module Enumerable
+  def my_each
+    i = 0
+    while i < length
+      yield self[i] if block_given?
+      i += 1
+    end
+    self
+  end
```
Comment (stratospherique) [inline enumerable_methods.rb]: * [ ] my_each returns an Enumerator if no block is given 
* [ ] my_each_with_index returns an enumerator if no block is given

File: enumerable_methods.rb [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+    end
+  end
+
+  def my_select
+    result = []
+    if block_given?
+      my_each { |i| result << i if yield(i) }
+    end
+    result
+  end
```
Comment (stratospherique) [inline enumerable_methods.rb]: * [ ] my_select returns an enumerator if no block is given

File: enumerable_methods.rb [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+      if block_given?
+        all = false if yield(i) == false
+      elsif patern == true
+        all = false if patern == i
+      else
+        all = false unless i != false
+      end
+    end
+    all
+  end
```
Comment (stratospherique) [inline enumerable_methods.rb]: * [ ] my_all? when a class is passed as an argument returns true if all of the collection is a member of such class
* [x] my_all? when a Regex is passed as an argument returns true if all of the collection matches the Regex
* [x] my_all? when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern

File: enumerable_methods.rb [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+      if block_given?
+        any = true if yield(i)
+      elsif patern == true
+        any = false if (patern == i) == false
+      else
+        any = false unless i != false
+      end
+    end
+    any
+  end
```
Comment (stratospherique) [inline enumerable_methods.rb]: * [ ] my_any? when no block or argument is given returns true if at least one of the collection is not false or nil
* [ ] my_any? when a class is passed as an argument returns true if at least one of the collection is a member of such class
* [ ] my_any? when a Regex is passed as an argument returns false if none of the collection matches the Regex
* [ ] my_any? when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern

File: enumerable_methods.rb [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+      if block_given?
+        result = false if yield(i)
+      elsif partern == true
+        result = false if (partern == i) == false
+      else
+        result = false unless i != false
+      end
+    end
+    result
+  end
```
Comment (stratospherique) [inline enumerable_methods.rb]: * [ ] my_none? when no block or argument is given returns true only if none of the collection members is true
* [ ] my_none? when a class is passed as an argument returns true if none of the collection is a member of such class
* [ ] my_none? when a Regex is passed as an argument returns true only if none of the collection matches the Regex
* [ ] my_none? when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern

File: enumerable_methods.rb [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+
+  def my_count
+    i = 0
+    if block_given?
+      my_each { i += 1 }
+    else
+      i = length
+    end
+    i
+  end
```
Comment (stratospherique) [inline enumerable_methods.rb]: * [ ] my_count counts the number of items in enum that are equal to item if an argument is given
* [ ] my_count counts the number of elements yielding a true value if a block is given

File: enumerable_methods.rb [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+
+  def my_map(&my_proc)
+    result = []
+    if block_given?
+      my_each { |element| result << my_proc.call(element) }
+    else
+      result = self
+    end
+    result
+  end
```
Comment (stratospherique) [inline enumerable_methods.rb]: * [ ] my_map returns an Enumerator if no block is given

File: enumerable_methods.rb [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+
+  def my_inject(memo = nil)
+    memo ||= self[0]
+    my_each { |val| memo = yield memo, val } if block_given?
+    memo
+  end
```
Comment (stratospherique) [inline enumerable_methods.rb]: * [ ] my_inject combines all elements of enum by applying a binary operation, specified by a block
* [ ] my_inject when a symbol is specified combines each element of the collection by applying the symbol as a named method

File: enumerable_methods.rb [hunk 18]
```diff
@@ -0,0 +1,117 @@
...
+print select
+all = numb.my_all? { |i| i < 8 }
+puts all
+any = name.my_any? { |i| i.is_a? Integer }
+puts any
+none = numb.my_none? { |i| i == 10 }
+print none
+puts numb.my_count
+print name.my_map(&my_proc)
+puts multiply_els(numb)
```
Comment (stratospherique) [inline enumerable_methods.rb]: - This code is irrelevant to the logic of enumerable methods and it should be removed.

**Review by stratospherique — CHANGES_REQUESTED** [review]
### Great work @collinsugwu  :clap: . However, some changes needs to be made in order for us to approve your project. check the code comments below :arrow_down: 

### That's it :relaxed: . When finished, kindly resubmit another code review request :pray: 

**Ahmed Mahfoudh**

File: enumerable_methods.rb (hunk 12 above)
Comment (collinsugwu) [reply enumerable_methods.rb]: please can you explain the points I checked a little more because I felt I solved that

**Review by uche-inyama — CHANGES_REQUESTED** [review]
## Hi Collins!
How are doing today :slightly_smiling_face: 

After running the test cases, it still fails the tests, 15 failures. 
              
              Finished in 0.03687 seconds (files took 0.12234 seconds to load) 28 examples, 15 failures


Please, go through this [Link](https://github.com/igakigongo/EnumerableMethods/blob/master/enumerable.rb), understand it and use it as a guide to update your project.

After implementing this change, submit another form for a code review.

#### ``` Have a great day ``` :slightly_smiling_face: 

``` Uchechukwu Henry Inyama ```

File: enumerable_methods.rb [hunk 19]
```diff
@@ -0,0 +1,90 @@
...
+
+  def my_count
+    i = 0
+    if block_given?
+      my_each { i += 1 }
+    else
+      i = length
+    end
+    i
+  end
```
Comment (maelfosso) [inline enumerable_methods.rb]:
```ruby
irb(main):011:0> array
=> [8, 0, 1, 8, 8, 0, 2, 6, 3, 1, 6, 8, 1, 7, 0, 6, 6, 1, 6, 1, 6, 6, 4, 6, 3, 2, 1, 3, 8, 3, 1, 3, 4, 7, 1, 0, 1, 1, 7, 8, 2, 5, 4, 7, 2, 5, 3, 1, 0, 8, 4, 8, 2, 0, 8, 0, 1, 3, 7, 6, 3, 7, 7, 0, 0, 3, 7, 2, 6, 7, 7, 2, 7, 8, 1, 3, 2, 6, 1, 0, 0, 3, 2, 7, 1, 5, 7, 6, 2, 6, 3, 8, 2, 8, 7, 0, 1, 4, 0, 8]
irb(main):012:0> array.count(0)
=> 13
irb(main):013:0> array.my_count(0)
Traceback (most recent call last):
        3: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        2: from (irb):13
        1: from enumerable_methods.rb:63:in `my_count'
ArgumentError (wrong number of arguments (given 1, expected 0))
irb(main):014:0> 
```

Comment (maelfosso) [reply enumerable_methods.rb]:
```ruby
irb(main):016:0> block = proc { |num| num < 9/2 }
=> #<Proc:0x0000557eb8d81940@(irb):16>
irb(main):017:0> array
=> [8, 0, 1, 8, 8, 0, 2, 6, 3, 1, 6, 8, 1, 7, 0, 6, 6, 1, 6, 1, 6, 6, 4, 6, 3, 2, 1, 3, 8, 3, 1, 3, 4, 7, 1, 0, 1, 1, 7, 8, 2, 5, 4, 7, 2, 5, 3, 1, 0, 8, 4, 8, 2, 0, 8, 0, 1, 3, 7, 6, 3, 7, 7, 0, 0, 3, 7, 2, 6, 7, 7, 2, 7, 8, 1, 3, 2, 6, 1, 0, 0, 3, 2, 7, 1, 5, 7, 6, 2, 6, 3, 8, 2, 8, 7, 0, 1, 4, 0, 8]
irb(main):018:0> array.count(&block)
=> 52
irb(main):019:0> array.my_count(&block)
=> 100
irb(main):020:0> 
```

File: enumerable_methods.rb [hunk 20]
```diff
@@ -0,0 +1,90 @@
...
+    result = []
+    my_each { |element| result << yield(element) }
+    result
+  end
+
+  def my_inject(memo = nil)
+    memo ||= self[0]
+    my_each { |val| memo = yield memo, val } if block_given?
+    memo
+  end
```
Comment (maelfosso) [inline enumerable_methods.rb]:
```ruby
irb(main):022:0> operation = proc { |sum, n| sum + n }
=> #<Proc:0x0000557eb8f5c828@(irb):22>
irb(main):023:0> array
=> [8, 0, 1, 8, 8, 0, 2, 6, 3, 1, 6, 8, 1, 7, 0, 6, 6, 1, 6, 1, 6, 6, 4, 6, 3, 2, 1, 3, 8, 3, 1, 3, 4, 7, 1, 0, 1, 1, 7, 8, 2, 5, 4, 7, 2, 5, 3, 1, 0, 8, 4, 8, 2, 0, 8, 0, 1, 3, 7, 6, 3, 7, 7, 0, 0, 3, 7, 2, 6, 7, 7, 2, 7, 8, 1, 3, 2, 6, 1, 0, 0, 3, 2, 7, 1, 5, 7, 6, 2, 6, 3, 8, 2, 8, 7, 0, 1, 4, 0, 8]
irb(main):024:0> array.inject(&operation)
=> 389
irb(main):025:0> array.my_inject(&operation)
=> 397
irb(main):026:0> 
```

Comment (maelfosso) [reply enumerable_methods.rb]:
```ruby
irb(main):028:0> array
=> [8, 0, 1, 8, 8, 0, 2, 6, 3, 1, 6, 8, 1, 7, 0, 6, 6, 1, 6, 1, 6, 6, 4, 6, 3, 2, 1, 3, 8, 3, 1, 3, 4, 7, 1, 0, 1, 1, 7, 8, 2, 5, 4, 7, 2, 5, 3, 1, 0, 8, 4, 8, 2, 0, 8, 0, 1, 3, 7, 6, 3, 7, 7, 0, 0, 3, 7, 2, 6, 7, 7, 2, 7, 8, 1, 3, 2, 6, 1, 0, 0, 3, 2, 7, 1, 5, 7, 6, 2, 6, 3, 8, 2, 8, 7, 0, 1, 4, 0, 8]
irb(main):029:0> array.inject(:+)
=> 389
irb(main):030:0> array.my_inject(:+)
=> :+
irb(main):031:0>
```

**Review by maelfosso — CHANGES_REQUESTED** [review]
Hello @collinsugwu

Really nice improvement in your code :clap: 
However, there are still some bugs you should fix to make the project approved.
Read the comments to know more about them.

Also, **Follow the Git Flow** otherwise the project won't be approved.
- Do not work directly in the master branch. Create the development branch and a specific feature-branch (give it a descriptive name) for every feature you implement. When your feature is ready and approved you can merge feature-branch to the development branch and when the release is ready create a pull request from development to master.
- Read more [here](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59882084/)

**Kindly make the changes and submit a new code review request**.
Thanks

**Mael FOSSO**
[Github](https://github.com) | [Twitter](https://twitter.com) | [LinkedIn](https://www.linkedin.com/in/mael-elvis-fosso-650b6346/)

File: enumerable_methods.rb [hunk 21]
```diff
@@ -78,10 +80,28 @@ def my_map
...
 
-  def my_inject(memo = nil)
-    memo ||= self[0]
-    my_each { |val| memo = yield memo, val } if block_given?
-    memo
+  def my_inject(*args)
+    arr = to_a.dup
+    if args[0].nil?
+      numn = arr.shift
+    elsif args[1].nil? && !block_given?
```
Comment (samgaco) [inline enumerable_methods.rb:87]: :100:

**Review by samgaco — APPROVED** [review]
Hi @collinsugwu :high_brightness: 

Finally done, congratulations. I hope you keep up the good work moving on. :+1: 

You can now go ahead and merge to the master branch and submit the project completion form.
--------------------------

* ###### STATUS OF THE PROJECT: APPROVED :heavy_check_mark:

**Pull request merged**